### ISSUE TITLE
A mixin class must have a constructor with a single rest parameter of type any[] => this is wrong

### ISSUE THREAD

**Issue by denis-migdal** [microsoft/TypeScript#59813]
### 🔎 Search Terms

None

### 🕗 Version & Regression Information

Tested on the nightly build.

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play/?ts=5.7.0-dev.20240831#code/C4TwDgpgBAwgzsATgHgCoD4oF4oG8B2EA7gBQB0FAhogOZwBcUl+IA2gLoCUjqAvgNwAoQQDMArvgDGwAJYB7fFAC2MgB4z8cNFAirgEfABM4sBCmYh06EgGsANpTgMoqTnmFRPURBGBjEipIOTlAAYnJyOnoGxlD2jia4Hl4pPpSGCnYgUCIRjPhiSgBGEIhCgikpkgpmYtJyiCS5cvmFJYgANFAUZNR0jBYcbkmVo1BwYpCNPX1wnEJjlcAAFjJwZM3YORELY7wVXvu8QA)

### 💻 Code

```ts
type Cstr<T> = {new(...args: any[]): T};

function mixins<T extends Cstr<any>>(klass: T) {

    return class Foo extends klass {

        readonly foo: number;

        constructor(foo: number, ...args: any[]) {
            super(...args);
            this.foo = foo;
        }
    }
}
```

### 🙁 Actual behavior

This raises the following error :
```
A mixin class must have a constructor with a single rest parameter of type any[]
```

### 🙂 Expected behavior

No TS errors should be raised.

### Additional information about the issue

Instead of asking for a single rest parameters, TS should ask the **last** parameters to be a rest parameters.

Also, when #59812, is fixed, it should also be possible to do more complex constructor signatures, without relying on `any[]` :

```ts
type Cstr<T> = {new(...args: any[]): T};

type InferCstrParams<T> = T extends {new(...args: infer U): T} ? U : never;

function mixins<T extends Cstr<any>>(klass: T) {

    return class Foo extends klass {

        readonly foo: number;

        constructor(foo: number, ...args: InferCstrParams<T>) {
            super(...args);
            this.foo = foo;
        }
    }
}
```

Indeed, the constraints should be on the `super()` call, NOT on the mixins constructor.

This would also enable to generate the constructor parameters when the mixins e.g. has a type restrictions on `klass`, e.g. :
```ts
type Cstr<T, U extends any[]> = {new(...args: U): T};
function mixins<T extends Cstr2<any,[number]>>(klass: T) {

    return class Foo extends klass {
        constructor(a: number, b: number) {
            super(a+b);
        }
    }


}
```

This would also enable to create mixins taking many classes as parameters, etc.

**Comment by jcalz** [comment]
You didn’t provide any search terms. Did you search for existing issues before filing this?
 If you search for `mixin constructor in:title` there are lots of hits. 
 
Duplicate

**Comment by denis-migdal** [comment]
From #14126

It seems a simple cast removes the error :
```ts
return class Foo extends (klass as { new(...args:[]): any }) {
    constructor(foo: number, b: boolean) { /* ... */ }
}
```

Note: if we want to add the base class parameters to the mixins class, we need to do another cast, cf the issue.

[Playground Link](https://www.typescriptlang.org/play/?ts=5.7.0-dev.20240831#code/C4TwDgpgBAwg9gWzHAzheA7FwBOBXAYx2DhwB4AVAGigFUA+KAXigCgoOoBtaugXSgQAHsAgYAJim4YIAdygAKAIYAuKAEsMAMwg4oAeQCMASmaNNOvQCVDVGfOVqLugwCZTTc9pdXXAgPxQAN7snGH2inBqRlAAZG7GajZxUL6hHAC+KQAK6gQA1pQ0+RAgcFpQFIzxuQVktMWl5XSMajIAbrrpbFp4GATA6nAYUABicHCUgiJiklARClFQShggHowrIPQKAEJKaGoUpiFhBAA2+1LjcNOiElK7+9D783KKAHSfqsurXHzrwQyx26p2G2HwA1IizAgzBaiCUC0EzU4M0AHMoEDgiCwpwUHhIDhobCsMYcbioAQwXAzhB3mc4GjiUMsO8kXAyRTMuTERMFMcMjzsEpBgReXB+YDuoKwjgIMA8DgRtdllJ4EhUOgwbhCMRSGRQJBmtcaFVWILWL1+iSoHtyBRbrMpAslpsAZttnsDpVgacLigpHbHfdFF7ns63o4fmszIDfVyqVgdZCiXAYSyUPCoAAjJQ4FG4dGY+NcvEE3TMsGc0ucRMoGl0hlMtMklDvXM4asUmVcjuSoI9inC0U5vP9weY1jdOUKpW2vPmqfnS7zgBewbm10eROXAeCxdM2NBSYhJFT6bh+-ZBZw6JoHZvd6gsgAFiKIJ181BURgMViTly+KEpWpI8nWDb0oyIFtq+76fl2k5hLmq7jt0w55KOKECoutLAJhzCvPIewoQiD5QAARORNDXhRVHPm+oifmolFAgA3KwyHtmOxjsZxyH8rxSirmyfJkqwxEiRKPHiUJXFEtJEn8TxQA)

Seriously, why are we forced to do such black magic found only in the darkness of the git issues, for such simple code ?
